libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/components/async/AsyncView.kt

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2023 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.libraries.designsystem.components.async
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.tooling.preview.PreviewParameter
+import io.element.android.libraries.architecture.Async
+import io.element.android.libraries.designsystem.components.ProgressDialog
+import io.element.android.libraries.designsystem.components.dialogs.ErrorDialog
+import io.element.android.libraries.designsystem.components.dialogs.ErrorDialogDefaults
+import io.element.android.libraries.designsystem.components.dialogs.RetryDialog
+import io.element.android.libraries.designsystem.preview.ElementPreview
+import io.element.android.libraries.designsystem.preview.PreviewsDayNight
+
+/**
+ * Render an Async object.
+ * - If Success, invoke the callback [onSuccess], only once.
+ * - If Failure, display a dialog with the error, which can be transformed, using [errorMessage]. When
+ * closed, [onErrorDismiss] will be invoked. If [onRetry] is not null, a retry button will be displayed.
+ * - When loading, display a loading dialog, if [showProgressDialog] is true, with on optional [progressText].
+ */
+@Composable
+fun <T> AsyncView(
+    async: Async<T>,
+    onSuccess: (T) -> Unit,
+    onErrorDismiss: () -> Unit,
+    modifier: Modifier = Modifier,
+    showProgressDialog: Boolean = true,
```

Comment:
Maybe we could use some defaults for this? Instead of directly using `showProgressDialog`, maybe have:

```kotlin
object AsyncViewDefaults {
    @Composable
    fun ProgressDialog(progressText: String? = null) { ... }

    @Composable
    fun SuccessCallback<T>(callback: (data: T) -> Unit) { 
        LaunchedEffect(Unit) { callback(data) } 
    }
}
```

Then this would be:

```kotlin
fun <T> AsyncView(
    ....
    onProgress: @Composable () -> Unit = { AsyncViewDefaults.ProgressDialog() },
    onSuccess: @Composable (Async.Success<T>) -> Unit = {},
)
```

And could be called as:

```kotlin
AsyncView(
    async = Async<Int>,
    ....
    onProgress = { AsyncViewDefaults.ProgressDialog() }, // Not really needed since it's default, just to showcase usage
    onSuccess = AsyncViewDefaults.SuccessCallback { data ->
        someCallback(data)
    },
)
```

And maybe the same for other `on...` parameters?

Reply:
That's sound like a good improvement of the API. I was not sure how to properly have default. Your proposal makes it clearer I think, I will see how it look in the code.

Reply:
Doing it for the ProgressDialog, I end up with this:

<img width="1672" alt="image" src="https://github.com/vector-im/element-x-android/assets/3940906/9f9c5f01-a3a4-4f21-b7e5-1eefc0b58056">

I believe for this particular case (since there will be no particular UI change) it's better to keep a short code at call site, so keep the existing API. We may iterate if the future prove that I was wrong.

Also for `onSuccess`, I did not manage to find a good solution, and I do not want to spend too much time on it. The existing API again is enought I think. Anyway thanks for the proposal!

I will rebase and generate the screenshot again.

Reply:
I will try to have a complete API and keep the existing code as a shortcut.